apiserver/common/crossmodel/auth.go

```diff
@@ -175,38 +170,40 @@ func (a *AuthContext) CheckLocalAccessRequest(details *offerPermissionCheck) ([]
 	return firstPartyCaveats, nil
 }
 
-func (a *AuthContext) checkOfferAccess(username, offerUUID string) error {
+type userAccessFunc func(names.UserTag, names.Tag) (permission.Access, error)
+
+func (a *AuthContext) checkOfferAccess(userAccess userAccessFunc, username, offerUUID string) error {
 	userTag := names.NewUserTag(username)
-	isAdmin, err := a.hasAdminAccess(userTag, permission.SuperuserAccess, a.systemState.ControllerTag())
-	if err != nil {
+	isAdmin, err := a.hasAccess(userAccess, userTag, permission.SuperuserAccess, a.systemState.ControllerTag())
+	if err != nil && !errors.Is(err, &apiservererrors.AccessRequiredError{}) {
 		return apiservererrors.ErrPerm
 	}
 	if isAdmin {
 		return nil
 	}
-	isAdmin, err = a.hasAdminAccess(userTag, permission.AdminAccess, a.systemState.ModelTag())
-	if err != nil {
+	isAdmin, err = a.hasAccess(userAccess, userTag, permission.AdminAccess, a.systemState.ModelTag())
+	if err != nil && !errors.Is(err, &apiservererrors.AccessRequiredError{}) {
 		return apiservererrors.ErrPerm
 	}
 	if isAdmin {
 		return nil
 	}
-	access, err := a.systemState.GetOfferAccess(offerUUID, userTag)
-	if err != nil && !errors.IsNotFound(err) {
-		return apiservererrors.ErrPerm
+	ok, err := a.hasAccess(userAccess, userTag, permission.ConsumeAccess, names.NewApplicationOfferTag(offerUUID))
```

Comment:
do the previous apiservererrors.AccessRequiredError errors also need to be handled?

Reply:
in the previous checks, AccessRequiredError just means no permission so check the next one
the for the last one, we want to return it if we have it, or if some other error, just the generic permission denied